spec.html

```diff
@@ -32585,12 +32508,90 @@ <h1>SortCompare ( _x_, _y_ )</h1>
             1. Let _ySmaller_ be the result of performing Abstract Relational Comparison _yString_ &lt; _xString_.
             1. If _ySmaller_ is *true*, return *1*<sub>𝔽</sub>.
             1. Return *+0*<sub>𝔽</sub>.
+          1. Return ? SortArrayLike(_obj_, _len_, _SortCompare_).
+        </emu-alg>
+        <emu-note>
+          <p>Because non-existent property values always compare greater than *undefined* property values, and *undefined* always compares greater than any other value, *undefined* property values always sort to the end of the result, followed by non-existent property values.</p>
+        </emu-note>
+        <emu-note>
+          <p>_SortCompare_ is not a consistent comparator if any of the following conditions is true:</p>
```

Comment:
Nit: this reads like it's a definition. I'd prefer something which makes it clearer that it is just an observation, such as

> It follows from the definition below that _SortCompare_ is consistent comparator unless one or more of the following conditions are met:

Reply:
re wording it as an observation: Yup, good point.

re changing it from "not consistent if" to "consistent unless": That presumes that the list is exhaustive, that there's no *other* way in which `_SortCompare_` could be not consistent. I *think* that's true, but I'm not 100% sure.

Reply:
On further reflection, I'm not convinced this refactoring (of the conditions necessary to trigger implementation-defined sort order, specifically) is an improvement. The original was clear that if _comparefn_ was undefined and any two calls to ToString on some value in `S` failed to return the same value then the sort order was implementation-defined. I can't tell whether that's entailed by the wording in this PR; if it is, it is much harder to see it.

(The only condition which appears related is reflexivity, but the way I read the definition of "consistent comparator", if it so happened that each time an implementation compared a value to itself it got `0` then the implementation would find reflexivity to hold, even if the observed ToString of that value varied on other occasions.)

Reply:
I was puzzled by this comment, but I think I figured out your point. 

For example, consider sorting 3 objects which I'll refer to as A, B, and C. `_comparefn_` is `*undefined*`, so SortCompare will use calls to ToString:
- ToString(A) always returns "a",
- ToString(B) always returns "b", and
- ToString(C) returns either "c1" or "c2". (But C is clever enough to always have ToString return "c1" if SortCompare is ever comparing C to itself.) 

So in both the status quo and this PR, the resulting SortCompare satisfies the requirements of a consistent comparison function (comparator). E.g., SortCompare(B,C) is always -1 regardless of what ToString(C) returns, because "b" < "c1" and "b" < "c2" are both true.

However, in the status quo, the inconsistent returns from ToString mean that the sort order is implementation-defined, whereas in this PR, the fact that _SortCompare_ is a consistent comparator is enough to make the sort order *not* implementation-defined.

So: in some cases where the status quo says the sort order is imp-def, this PR says that the sort order is *not* imp-def. Is that your point?

If so, ...  I don't know, it seems to me that being a consistent comparison function should be the only requirement. As long as C's @@toPrimitive method can make SortCompare behave as a consistent comparison function, then it shouldn't matter if it's doing weird things at a level that SortArrayLike can't see. But yeah, I guess that's a normative change.

Reply:
> So: in some cases where the status quo says the sort order is imp-def, this PR says that the sort order is not imp-def. Is that your point?

Right. Sorry that wasn't clearer.

> I don't know, it seems to me that being a consistent comparison function should be the only requirement.

I can see that argument, but I wouldn't want to push for that normative change - it seems reasonable-ish for an implementation to bail as soon as it sees inconsistent ToString results, even if they're inconsistent in a way which doesn't matter; having to figure out _if_ it matters could be an additional burden. And there's no strong reason to want to add strict requirements for sort order when sorting sets which have elements with non-constant ToString behavior.

Reply:
> ... it seems reasonable-ish for an implementation to bail as soon as it sees inconsistent ToString results, even if they're inconsistent in a way which doesn't matter; having to figure out _if_ it matters could be an additional burden. ...

But an implementation *doesn't* have to figure out if it matters, because it doesn't have to *detect* when the sort order is imp-def. It just has to have code that will work in the non-imp-def cases; what that code does in imp-def cases doesn't matter.

This PR is saying: you (the implementation) have to prove to yourself that your `sort` code works as long as SortCompare is consistent.
The status quo says: you have to prove it works in cases where SortCompare is consistent *minus* those cases where SortCompare is consistent but ToString is inconsistent. But I don't think that "minus" helps: you'll just prove it works for all consistent-SortCompare cases and ignore that that covers some cases you didn't need to cover.

Reply:
> But I don't think that "minus" helps: you'll just prove it works for all consistent-SortCompare cases and ignore that that covers some cases you didn't need to cover.

Eh. I can imagine an implementation which keeps track of whether ToString is consistent and bails if not, rather than writing code that blindly ignores it. Some implementations are more aggressive about checking internal invariants than others.

In any case, it sounds like we're agreed this would be a normative change, so it would need to go in its own PR either way. (Separately, I doubt there's much appetite for it.)